core/src/main/scala/kafka/log/Log.scala

```diff
@@ -677,6 +672,47 @@ class Log(val dir: File,
   }
 
   /**
+    * Find any segments that match the user-supplied predicate UNLESS it is the final segment
+    * and it is empty (since we would just end up re-creating it)
+    * @param predicate A function that takes in a single log segment and returns true iff it is deletable
+    * @return the segments ready to be deleted
+    */
+  private def deletableSegments(predicate: LogSegment => Boolean) = {
+    val lastEntry = segments.lastEntry
+    if (lastEntry == null) Seq.empty
+    else logSegments.takeWhile(s => predicate(s) && (s.baseOffset != lastEntry.getValue.baseOffset || s.size > 0))
+  }
+
+  /**
+    * Delete any log segments that have either expired due to time based retention
+    * or because the log size is > retentionSize
+    */
+  def deleteOldSegments(): Int = {
+    if (!config.delete) return 0
+    deleteRetenionMsBreachedSegments() + deleteRetentionSizeBreachedSegments()
+  }
+
+  private def deleteRetenionMsBreachedSegments() : Int = {
```

Comment:
Yeah its not major thing. So I am okay either way. 

The added benefit I could see it that the predicates could be combined in cases where both were being applied. This would prevent iterating over the segments twice when evaluating them to be deleted. I am not sure how much impact that would actually have though.